Added selected styling to custom child in Select

jcfilben opened the pull request:
<!--- Provide a general summary of the PR in the Title above -->

#### What does this PR do?
When you have custom children in Select the hover styling is applied but the selected styling is not applied. This PR adds the selected styling to custom children when they are selected.

#### Where should the reviewer start?
SelectContainer.js

#### What testing has been done on this PR?
Tested with this storybook story
```javascript
import React from 'react';

import { Box, Grommet, Select, Text, Paragraph } from 'grommet';
import { grommet } from 'grommet/themes';

export const TEST = () => {
  const options = ['one', 'two'];
  return (
    <Grommet full theme={grommet}>
        <Box gap="small" width="small">
            <Select options={options} />
            <Select options={options}>{(option) => <Text>{option}</Text>}</Select>
            <Select options={options}>{(option) => option}</Select>
            <Select options={options}>{(option) => <Paragraph>{option}</Paragraph>}</Select>
        </Box>
    </Grommet>
  );
};

export default {
  title: 'Input/Select/TEST',
};
```

#### How should this be manually tested?
See above

#### Any background context you want to provide?

#### What are the relevant issues?
Closes #5154 

#### Screenshots (if appropriate)

#### Do the grommet docs need to be updated?
No

#### Should this PR be mentioned in the release notes?
Yes

#### Is this change backwards compatible or is it a breaking change?
Backwards compatible

ericsoderberghp commented on src/js/components/Select/SelectContainer.js:
```diff
@@ -382,11 +382,15 @@ const SelectContainer = forwardRef(
-                      selected: optionSelected,
-                    });
+                    child = (
+                      <OptionBox selected={optionSelected}>
```
Should we also be passing `{...selectOptionsStyle}`?

jcfilben replied:
I don't think we should pass `{...selectOptionsStyle}` because this adds padding to each option and if the user is passing a custom child I don't think grommet should automatically add padding on it.

ericsoderberghp replied:
Hmm... tricky to know what the theme might have in `theme.select.options.container`. Agreed on padding being a problem. Not so sure though if it has border or background. I think if we add `selected` to `Button`, this aspect isn't an issue.

jcfilben replied:
`theme.select.options.container` has the following properties set: 
```
options: {
        container: {
          align: 'start',
          pad: 'small',
        },
```
But as you said, if we are doing the selected styling in the Button then this won't matter

ericsoderberghp replied:
I was thinking we just add `${props => props.selected && selectedStyle}` to Button. That would leave the select options container styling to always be supplied unless the caller passes a children render function, keeping the `plain` nature of that path.

ericsoderberghp commented on src/js/components/Select/SelectContainer.js:
```diff
@@ -382,11 +382,15 @@ const SelectContainer = forwardRef(
+                    );
                   else if (theme.select.options)
                     child = (
                       <OptionBox
```
Wondering if we should reduce duplication and do `{child && <OptionBox ...>{child}</OptionBox>}` inside `<SelectOption />`?

jcfilben replied:
Since we are passing different props `selected={optionSelected}` in the first case and `selected={optionSelected}` and `{...selectOptionsStyle}` in the second case is it possible to reduce duplication?

ericsoderberghp replied:
not in a significantly cleaner way

ericsoderberghp commented on src/js/components/Select/SelectContainer.js:
```diff
@@ -398,6 +402,7 @@ const SelectContainer = forwardRef(
+
                   // if we have a child, turn on plain, and hoverIndicator
 
                   return (
```
Would it make any sense to add `selected` as a prop to `Button` and then do that styling, in the same we do `active` styling? I'm not sure, so just wondering.

jcfilben replied:
If the styling for a selected option is done in Button it changes the behavior when a selected option is hovered over the background color changes to the `active` color. Current behavior for Select is that when a selected option is hovered over the background color doesn't change.

ericsoderberghp replied:
I'm wondering if we should reconsider that Button behavior. Is there anywhere else where we use `<Button selected />`? I noticed that there is currently no `selected` property documented for Button. I was wondering if we should add it, and have its behavior be what you mention, that hover doesn't affect its styling.

jcfilben replied:
I didn't find other places where `<Button selected />` is used. The places where I thought it might come up (like on Calendar) are not doing the selected styling on the button, similar to how Select is currently set up.

It makes sense to me to change the selected behavior on button so that the hover isn't changing the background. I can go ahead and do that and add the selected prop to the documentation.

ericsoderberghp replied:
OK. I think we might consider not adding it to the documentation to start with, and just consider this an internal usage for now. Then, if demand arises, we can add it. I say this in the spirit of not growing our published API surface unless there are concrete needs for it.

jcfilben commented:
I changed Button so that it is handling the selected styling. I realized that in storybook we have stories where when an option is selected and the selection is indicated in a different way than the background color. For example the Many Options story.

**Here is the Many Options story with `selected` styling being handled by Button:**
<img width="314" alt="Screen Shot 2021-06-03 at 3 58 41 PM" src="https://user-images.githubusercontent.com/54560994/120717479-d993c780-c484-11eb-98bd-e68b4f3e7fc6.png">
**Here is the Many Options story with `selected` styling being handled by Select:**
<img width="319" alt="Screen Shot 2021-06-03 at 3 59 02 PM" src="https://user-images.githubusercontent.com/54560994/120717488-dc8eb800-c484-11eb-8ef9-a671cee78bee.png">

If button is handling the selected styling it will always force the background color to change

ericsoderberghp commented:
Agreed on wanting the children renderer to be able to control the select styling, as shown in the Many Options example with CheckBoxes. But, I see the same problem you showed for the Button approach with the approach in this PR: https://deploy-preview-5296--sad-tereshkova-173d07.netlify.app/?path=/story/input-select-many-options--many-options

I'm wondering if we should tweak the approach to say that if the children function returns a string, we will include the select styling, but otherwise we will leave it as is. I'm trying to think of another approach to allow both types of behavior, but haven't thought of something clean yet. Any suggestions?

jcfilben commented:
> I see the same problem you showed for the Button approach with the approach in this PR

Good catch I didn't notice that before. In the latest commit I updated Select to handle selected styling through Button and I fixed the issue with the Many Options story by only applying selected styling to custom children if they only contain text.

ericsoderberghp reviewed (approved):
Approved, but I think we should give this a little time on `stable` before releasing, so I'm delaying the merge until after the pending release.
